virtio-devices/src/watchdog.rs

```diff
@@ -0,0 +1,432 @@
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+//
+// Copyright © 2020 Intel Corporation
+// Portions Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE-BSD-3-Clause file.
+
+use super::Error as DeviceError;
+use super::{
+    ActivateError, ActivateResult, EpollHelper, EpollHelperError, EpollHelperHandler, Queue,
+    VirtioCommon, VirtioDevice, VirtioDeviceType, EPOLL_HELPER_EVENT_LAST, VIRTIO_F_VERSION_1,
+};
+use crate::seccomp_filters::{get_seccomp_filter, Thread};
+use crate::{VirtioInterrupt, VirtioInterruptType};
+use anyhow::anyhow;
+use seccomp::{SeccompAction, SeccompFilter};
+use std::fs::File;
+use std::io::{self, Read};
+use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};
+use std::result;
+use std::sync::atomic::AtomicBool;
+use std::sync::{Arc, Barrier, Mutex};
+use std::thread;
+use std::time::Instant;
+use vm_memory::{Bytes, GuestAddressSpace, GuestMemoryAtomic, GuestMemoryMmap};
+use vm_migration::{
+    Migratable, MigratableError, Pausable, Snapshot, SnapshotDataSection, Snapshottable,
+    Transportable,
+};
+use vmm_sys_util::eventfd::EventFd;
+
+const QUEUE_SIZE: u16 = 8;
+const QUEUE_SIZES: &[u16] = &[QUEUE_SIZE];
+
+// New descriptors are pending on the virtio queue.
+const QUEUE_AVAIL_EVENT: u16 = EPOLL_HELPER_EVENT_LAST + 1;
+// Timer expired
+const TIMER_EXPIRED_EVENT: u16 = EPOLL_HELPER_EVENT_LAST + 2;
+
+// Number of seconds to check to see if there has been a ping
+// This needs to match what the driver is using.
+const WATCHDOG_TIMER_INTERVAL: i64 = 15;
+
+// Number of seconds since last ping to trigger reboot
+const WATCHDOG_TIMEOUT: u64 = WATCHDOG_TIMER_INTERVAL as u64 + 5;
+
+struct WatchdogEpollHandler {
+    queues: Vec<Queue>,
+    mem: GuestMemoryAtomic<GuestMemoryMmap>,
+    interrupt_cb: Arc<dyn VirtioInterrupt>,
+    queue_evt: EventFd,
+    kill_evt: EventFd,
+    pause_evt: EventFd,
+    timer: File,
+    last_ping_time: Arc<Mutex<Option<Instant>>>,
```

Comment:
One suggestion: Instead of timestamping the last ping, we could initially set the timer up and then resetting it (disarm,arm) when receiving a ping. A timer expiration would mean we need to reset, there would no longer need to track the ping time and computed elapsed times since last ping in the timer expiration handler.

Reply:
That is an interesting suggestion thank you @sameo. Currently it takes advantage of the periodic timer so it is not necessary to teardown/setup the timer each time which I feel is a bit cleaenr. It also took a bit of effort to make sure everything worked correctly across pause/resume (and by extension snapshot/restore) so i'm loath to fiddle with the state mechanics now that I know that is working correctly!